### PR TITLE
Avoid overflow in PaymasterERC20._erc20Cost rounding (L-15)

### PR DESCRIPTION
**Pull request by Amxx**
## Fixes L-15

`PaymasterERC20._erc20Cost` guards against overflow of the floor division with the `high < denominator` check, but the previous implementation performed the round-up via `Math.mulDiv(..., Math.Rounding.Ceil)`.

In the edge case where the floor result is exactly `type(uint256).max` and the division has a nonzero remainder, the internal `+1` of `Rounding.Ceil` overflows and reverts, instead of returning the intended saturated value.

This change computes the floor division and applies the round-up bit explicitly via `saturatingAdd`, so the function saturates at `type(uint256).max` rather than reverting.

```solidity
high < denominator
    ? nativeCost.mulDiv(tokenPerNative, denominator).saturatingAdd(
        (mulmod(nativeCost, tokenPerNative, denominator) > 0).toUint()
    )
    : type(uint256).max;
```

The rounding semantics are unchanged for all non-saturating inputs.

No changeset: behavior change is limited to an unreleased contract edge case.

### CODE REVIEW

File: test/account/paymaster/PaymasterERC20.test.js [hunk 1]
```diff
@@ -323,4 +323,24 @@ describe('PaymasterERC20', function () {
...
+      const tokenPerNative = ethers.MaxUint256;
+      const nativeCost = ethers.MaxUint256;
+
+      await expect(this.paymaster.$_erc20Cost(nativeCost, tokenPerNative)).to.eventually.equal(ethers.MaxUint256);
+    });
+
+    it('_erc20Cost rounds up without overflowing when the ceil result saturates', async function () {
+      // Regression (L-15): with tokenPerNative = denominator + 1, this nativeCost makes the floor division
+      // land exactly on type(uint256).max with a non-zero remainder. Rounding up must saturate rather than
+      // overflow (the previous mulDiv(..., Ceil) implementation reverted here).
```
Comment (luiz-lvj) [inline test/account/paymaster/PaymasterERC20.test.js]: I think we should not reference the issue in the code
```suggestion
      // with tokenPerNative = denominator + 1, this nativeCost makes the floor division
      // land exactly on type(uint256).max with a non-zero remainder. Rounding up must saturate rather than
      // overflow (the previous mulDiv(..., Ceil) implementation reverted here).
```

**Review by luiz-lvj — COMMENTED** [review]
just a nit

File: test/account/paymaster/PaymasterERC20.test.js (hunk 1 above)
Comment (Amxx) [reply test/account/paymaster/PaymasterERC20.test.js]:
```suggestion
      // Values provided here makes the floor division land exactly on type(uint256).max with a non-zero remainder. 
      // We check that the saturating addition that implements the rounding up does not overflow like the default
      // `Math.mulDiv(..., Math.Rounding.Ceil)` would.
```

File: test/account/paymaster/PaymasterERC20.test.js [hunk 2]
```diff
@@ -323,4 +323,24 @@ describe('PaymasterERC20', function () {
...
+  describe('edge cases', function () {
+    it('_erc20Cost returns max uint256 without reverting when muldiv overflows', async function () {
+      const tokenPerNative = ethers.MaxUint256;
+      const nativeCost = ethers.MaxUint256;
+
+      await expect(this.paymaster.$_erc20Cost(nativeCost, tokenPerNative)).to.eventually.equal(ethers.MaxUint256);
+    });
+
+    it('_erc20Cost rounds up without overflowing when the ceil result saturates', async function () {
+      // Values provided here makes the floor division land exactly on type(uint256).max with a non-zero remainder.
```
Comment (luiz-lvj) [inline test/account/paymaster/PaymasterERC20.test.js]:
```suggestion
      // Values provided here make the floor division land exactly on type(uint256).max with a non-zero remainder.
```

File: test/account/paymaster/PaymasterERC20.test.js [hunk 3]
```diff
@@ -323,4 +323,24 @@ describe('PaymasterERC20', function () {
...
```
Comment (Amxx) [inline test/account/paymaster/PaymasterERC20.test.js]:
```suggestion
      // Values provided here make the floor division land exactly on type(uint256).max with a non-zero remainder.
```

**Review by luiz-lvj — APPROVED** [review]
lgtm

**Review by james-toussaint — APPROVED** [review]
Ok, close to `Math.mulDiv(x, y, d, r)` but saturating instead of reverting on ceiling.
